### PR TITLE
Store lexer error offset relative to leading trivia start

### PR DESCRIPTION
**Pull request by ahoppen**
We need this so we can diagnose invalid UTF-8 in a token’s leading trivia. Also, it seems cleaner in general to store the lexer error relative to the token’s leading trivia start instead of the token’s text start.

### CODE REVIEW

**Comment by ahoppen** [comment]
@swift-ci Please test

File: Sources/SwiftSyntax/Raw/RawSyntaxTokenView.swift [hunk 1]
```diff
@@ -65,6 +65,29 @@ public struct RawSyntaxTokenView {
     }
   }
 
+  @_spi(RawSyntax)
+  public func wholeText<T>(_ body: (SyntaxText) -> T) -> T {
```
Comment (bnbarham) [inline Sources/SwiftSyntax/Raw/RawSyntaxTokenView.swift:69]: Could we instead have a function to get the `UnicodeScalar` for a particular offset?

**Review by bnbarham — APPROVED** [review]

File: Sources/SwiftSyntax/Raw/RawSyntaxTokenView.swift (hunk 1 above)
Comment (ahoppen) [reply Sources/SwiftSyntax/Raw/RawSyntaxTokenView.swift:69]: The current diagnostics only reference a single character but I could imagine that future diagnostics might reference an entire substring, e.g. from the offset to the end of the token, so I think the added flexibility of this API is good.

**Comment by ahoppen** [comment]
@swift-ci Please test macOS

File: Sources/SwiftSyntax/Raw/RawSyntaxTokenView.swift (hunk 1 above)
Comment (bnbarham) [reply Sources/SwiftSyntax/Raw/RawSyntaxTokenView.swift:69]: That's true... in that case Doug added a `withEachSyntaxText` that you could use instead. It's a ... little more involved, but would avoid the temporary buffer creation (well, we still make it for leading/trailing).

Alternatively given this is only ever used in the error case, you could just use `syntaxTextBytes` (which uses `withEachSyntaxText` but writes it into an array).

This function is just sort of half way, which I don't really see the point of. Either we care about performance, in which case we should use `withEachSyntaxText`. Or we don't and thus `syntaxTextBytes` is simpler.

Comment (ahoppen) [reply Sources/SwiftSyntax/Raw/RawSyntaxTokenView.swift]: Using `syntaxTextBytes` sounds good to me.

**Comment by ahoppen** [comment]
@swift-ci Please test

File: Sources/SwiftParserDiagnostics/LexerDiagnosticMessages.swift [hunk 2]
```diff
@@ -98,9 +98,11 @@ public extension SwiftSyntax.LexerError {
   /// `tokenText` is the entire text of the token in which the `LexerError`
   /// occurred, including trivia.
   @_spi(RawSyntax)
-  func diagnostic(wholeText: SyntaxText) -> DiagnosticMessage {
+  func diagnostic(wholeTextBytes: [UInt8]) -> DiagnosticMessage {
     var scalarAtErrorOffset: UnicodeScalar {
-      Unicode.Scalar(wholeText[Int(self.byteOffset)])
+      // Fall back to the Unicode replacement character U+FFFD in case we can't
+      // lex the unicode character at `byteOffset`. It's the best we can do
+      Unicode.Scalar.lexing(from: wholeTextBytes[Int(self.byteOffset)...]) ?? UnicodeScalar("�")
```
Comment (bnbarham) [inline Sources/SwiftParserDiagnostics/LexerDiagnosticMessages.swift:105]: The other option would be to display the hex value.

File: Sources/SwiftParser/Lexer/Cursor.swift [hunk 3]
```diff
@@ -674,62 +674,7 @@ extension Lexer.Cursor {
...
-    // Number of bits in the value, ignoring leading zeros.
-    let numBits = 32 - charValue.leadingZeroBitCount
-    if numBits <= 5 + 6 {
-      return encodedBytes == 2 ? Unicode.Scalar(charValue) : nil
-    }
-    if numBits <= 4 + 6 + 6 {
-      return encodedBytes == 3 ? Unicode.Scalar(charValue) : nil
-    }
-    return encodedBytes == 4 ? Unicode.Scalar(charValue) : nil
+    return Unicode.Scalar.lexing(advance: { self.advance() }, peek: { self.peek(at: 0) })
```
Comment (bnbarham) [inline Sources/SwiftParser/Lexer/Cursor.swift:677]: Any slowdown on this or is it optimized it fine?

Comment (ahoppen) [reply Sources/SwiftParser/Lexer/Cursor.swift:677]: Thanks for asking. I needed to add another `@inline(__always)` to fix a performance regression.

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
@swift-ci Please test

File: Sources/SwiftParserDiagnostics/LexerDiagnosticMessages.swift (hunk 2 above)
Comment (ahoppen) [reply Sources/SwiftParserDiagnostics/LexerDiagnosticMessages.swift:105]: Let’s stick with the replacement character. I don’t really expect anyone to hit this anyway and if they do, I don’t think a hex value would help them any more.

**Review by rintaro — APPROVED** [review]

**Pull request merged**